File: programmers/코딩테스트연습/그래프/네트워크.cpp
#include <string>
#include <vector>
#include <iostream>
#include <map>

using namespace std;
bool visited[200];
map<int, vector<int>> network;
void dfs(vector<vector<int>> &computers, int node)
{
    if (visited[node])
    {
        return;
    }
    visited[node] = true;
    for (int i = 0; i < network[node].size(); i++)
    {
        dfs(computers, network[node][i]);
    }
}
int solution(int n, vector<vector<int>> computers)
{
    int answer = 0;
    for (int i = 0; i < computers.size(); i++)
    {
        for (int j = 0; j < computers[0].size(); j++)
        {
            if (computers[i][j] == 1)
            {
                network[i].push_back(j);
            }
        }
    }
    for (int i = 0; i < computers.size(); i++)
    {
        if (visited[i])
        {
            continue;
        }
        dfs(computers, i);
        answer++;
    }
    return answer;
}

int main()
{
    vector<vector<int>> computers = {{1, 1, 0}, {1, 1, 1}, {0, 1, 1}};
    int n = 3;
    cout << solution(n, computers) << "\n";
}

File: programmers/코딩테스트연습/그래프/네트워크.py
def solution(n, computers):
    visited = [False for _ in range(n)]
    cnt = 0

    def dfs(i):
        visited[i] = True
        for j in range(n):
            if i == j:
                continue
            print(f"i: {i}, j: {j}")
            if computers[i][j] == 1 and not visited[j]:
                dfs(j)

    for i in range(n):
        if not visited[i]:
            dfs(i)
            cnt += 1
    return cnt


if __name__ == "__main__":
    print(solution(3, [[1, 1, 0], [1, 1, 0], [0, 0, 1]]))
